### PR TITLE
Fix FA2 long-context perf regression from removed .item() sync (#46693)

### LINKED ISSUE #46693: Performance regression with FA2 and long-context generation after 5.2.0
### System Info

transformers: 5.2.0 vs 5.4.0 (I could not test 5.3.0 due to a bug)
PyTorch: 2.12
flash-attn: 2.8.3
CUDA: 13.2
Python: 3.12
cudnn: 9.20.0.48

### Who can help?

@zucchini-nlp 
@Cyrilvallez 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

1. Install the latest version of transformers
2. Install FlashAttention2
3. Run the provided code
4. Downgrade to transformers version 5.2.0
5. Run the code again

```python
from transformers import Qwen3VLForConditionalGeneration, AutoProcessor
import torch

model = Qwen3VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen3-VL-2B-Thinking",
    dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",
).to("cuda")

processor = AutoProcessor.from_pretrained("Qwen/Qwen3-VL-2B-Thinking")

messages = [
    [
        {
            "role": "user",
            "content": [
                {"type": "video", "video": torch.zeros(768, 3, 512, 512)},
                {"type": "text", "text": "placeholder"},
            ],
        }
    ],
    [
        {
            "role": "user",
            "content": [
                {"type": "video", "video": torch.zeros(768, 3, 512, 512)},
                {"type": "text", "text": "different placeholder"},
            ],
        }
    ],
]

inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    do_sample_frames=False,
    return_dict=True,
    return_tensors="pt",
    padding=True,
    padding_side="left",
).to("cuda")

MAX_LEN = 16 * 1024

while True:
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)

    start.record()
    out_ids = model.generate(
        **inputs, do_sample=False, min_length=MAX_LEN, max_length=MAX_LEN, use_cache=True
    )
    end.record()

    torch.cuda.synchronize()
    assert out_ids.size(0) == 2 and out_ids.size(1) == MAX_LEN

    cur_time = start.elapsed_time(end)
    print(cur_time)
```

### Expected behavior

Tested on an NVIDIA RTX 6000 PRO:
- v5.2.0 FA2 = 92 sec
- v5.4.0 FA2 = 108 sec (+17% overhead)
- v5.2.0 SDPA = 136 sec
- v5.4.0 SDPA = 136 sec (no change)

The performance regression seems to occur when the KV cache is sufficiently large, at which point the GPU memory bandwidth has been saturated. Furthermore, it seems to be related with the changes to the FA2 integration layer. It does not seem to be related with the nvidia-cudnn conv kernel performance regression which was fixed in a recent PyTorch version.

### PR DESCRIPTION
**Pull request by ammmanism**
What does this PR do?
Fixes #46693 — FA2 long-context generation perf regression (5.2.0 → 5.4.0+, ~17% slower on large KV caches).
Supersedes #47215 (closed — that branch had picked up unrelated commits from other local branches, causing CI/security-gate noise unrelated to this fix). This PR is scoped to exactly 3 files.
Root cause
PR #42435 removed .item() calls on max_seqlen_in_batch in three places, leaving it as a 0-dim CUDA tensor instead of a Python int:

_get_unpad_data()
_unpad_input()
prepare_fa_kwargs_from_position_ids()

These tensors are passed directly into flash_attn_varlen_func(max_seqlen_q=..., max_seqlen_k=...). On bandwidth-saturated GPUs with large KV caches (RTX 6000 Pro noticeably more than RTX 4090), this adds kernel-dispatch/sync overhead per decode step that compounds over long generations.
Fix
Restored .item() at all three call sites, guarded by is_torchdynamo_compiling() so torch.compile graph capture isn't broken — tensor form preserved under compile, scalar form used in eager. Also updated _process_flash_attention_kwargs() to convert to int in eager mode consistently (previously only converted during tracing).
Verification

Regression test asserting scalar type in eager mode (tests/utils/test_flash_attention_utils.py)
Standalone microbenchmark isolating the mechanism (bench_fa2_unpad.py)
ruff check / ruff format --check clean, utils/check_copies.py passes

Relates to the investigation in #47134 (draft) — this should close the residual gap flagged there.
cc @vasqu @zucchini-nlp @andreasgoulas
Before submitting

 Did you read the contributor guideline and the Pull Request checks?
 Was this discussed/approved via a GitHub issue? Yes — #46693
 Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -673,15 +680,23 @@ def _process_flash_attention_kwargs(
...
 
     if supports_mapping["max_seqlen_k"] and max_seqlen_k is not None:
         if same_max_seqlen and flash_kwargs["max_seqlen_q"] is not None:
             max_seqlen_k = flash_kwargs["max_seqlen_q"]
-        elif not isinstance(max_seqlen_k, int) and is_tracing(max_seqlen_k):
-            max_seqlen_k = max_seqlen_k.item()
+        elif not isinstance(max_seqlen_k, int):
+            if is_torchdynamo_compiling():
+                # Under compile, keep as tensor to avoid graph break
+                pass
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py:697]: Yep no I think this is wrong, under compile we are forced to call item as FA fake ops are expecting a base int so we cannot avoid it

#47134 seems to fix the regression already so I don't see a reason to add this additional guarding instead

**Comment by ammmanism** [comment]
Thanks for the review and the correction on the compile-mode guard — makes sense, appreciate you catching that. Since #47134 is merged and already resolves the regression, closing this in favor of that fix.
